src/data/polynomial/denominators.lean

```diff
@@ -0,0 +1,90 @@
+/-
+Copyright (c) 2020 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import data.polynomial.erase_lead
+
+/-!
+# Denominators of evaluation of polynomials at ratios
+
+Let `i : R → K` be a homomorphism of semirings.  Assume that `K` is commutative.  If `a` and
+`b` are elements of `R` such that `i b ∈ K` is invertible, then for any polynomial
+`f ∈ polynomial R` the "mathematical" expression `b ^ f.nat_degree * f (a / b) ∈ K` is in
+the image of the homomorphism `i`.
+-/
+
+open polynomial finset
+
+/--  An induction lemma for polynomials. It takes a natural number `N` as a parameter, that is
+required to be at least as big as the `nat_degree` of the polynomial.  This is useful to prove
+results where you want to change each term in a polynomial to something else depending on the
+`nat_degree` of the polynomial itself and not on the specific `nat_degree` of each term. -/
+lemma induction_with_nat_degree_le {R : Type*} [semiring R] {P : polynomial R → Prop} (N : ℕ)
+  (P_0 : P 0)
+  (P_C_mul_pow : ∀ n : ℕ, ∀ r : R, r ≠ 0 → n ≤ N → P (C r * X ^ n))
+  (P_C_add : ∀ f g : polynomial R, f.nat_degree ≤ N → g.nat_degree ≤ N → P f → P g → P (f + g)) :
+  ∀ f : polynomial R, f.nat_degree ≤ N → P f :=
+begin
+  intros f df,
+  generalize' hd : card f.support = c,
+  revert f,
+  induction c with c hc,
+  { exact λ f df f0, by rwa (finsupp.support_eq_empty.mp (card_eq_zero.mp f0)) },
+  { intros f df f0,
+    rw ← erase_lead_add_C_mul_X_pow f,
+    refine P_C_add f.erase_lead _ (le_trans erase_lead_nat_degree_le df) _ _ _,
+    { exact le_trans (nat_degree_C_mul_X_pow_le (leading_coeff f) (nat_degree f)) df },
+    { exact hc _ (le_trans erase_lead_nat_degree_le df) (erase_lead_card_support f0) },
+    { refine P_C_mul_pow _ _ _ df,
+      rw [ne.def, leading_coeff_eq_zero],
+      rintro rfl,
+      exact (not_le.mpr (nat.succ_pos c)) (eq.ge f0) } },
+end
+
+section clearing_denominators
+
+variables {R K : Type*} [semiring R] [comm_semiring K] {i : R →+* K}
+variables {a b : R} {bi : K} {bu : bi * i b = 1}
```

Comment:
Do you actually use this `bu`? It seems that you add it as explicit assumption to every declaration that uses it.

Reply:
I removed bu as a variable, since I only used it when I wanted it explicit: thanks for catching this!